In [1]:
from math import log
import numpy as np

## Problem 1

### a)
Compute the P A = LU factorization of
$$A = \left[ \begin{array}{ccc} 4.0 & 2.0 & 1.0 \\ 9.0 & 3.0 & 1.0 \\ 25.0 & 5.0 & 1.0 \end{array} \right] $$
using maximal element partial pivoting. Compute all intermediate values as decimal approximations rounded to six significant digits – no fractions!

$$A = \left[ \begin{array}{ccc} 4.0 & 2.0 & 1.0 \\ 9.0 & 3.0 & 1.0 \\ 25.0 & 5.0 & 1.0 \end{array} \right] \sim \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 9.0 & 3.0 & 1.0 \\ 4.0 & 2.0 & 1.0 \end{array} \right] $$
Note that we changed rows 1 and 3
$$\sim \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 0 & 1.2 & 0.64 \\ 4.0 & 2.0 & 1.0 \end{array} \right]$$
This give a $l_{2,1} = .36$

$$\sim \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 0 & 1.2 & 0.64 \\ 0.0 & 1.2& 0.84 \end{array} \right]$$ 
This give a $l_{3,1} = .16$

$$\sim \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 0 & 1.2 & 0.64 \\ 0.0 & 1& 0.2 \end{array} \right]$$ 
This give a $l_{3,2} = 1.0$

Hence,
$$P = \left[ \begin{array}{ccc} 0.0 & 0.0 & 1.0 \\ 0.0 & 1.0 & 0.0 \\ 1.0 & 0.0 & 0.0 \end{array} \right]$$ 

$$L = \left[ \begin{array}{ccc} 1.0 & 0.0 & 0.0 \\ 0.36 & 1.0 & 0.0 \\ 0.16 & 1.0 & 1.0 \end{array} \right]$$ 

$$U = \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 0 & 1.2 & 0.64 \\ 0.0 & 0.0 & 0.2 \end{array} \right]$$ 






### b)
Use this $PA = LU$ factorization to solve
$$Ax=b=\left[ \begin{array}{c} 0.693147 \\ 1.098612 \\  1.386294 \end{array} \right] $$
Show intermediate steps, with values rounded to six decimal places.

Hence We first have $Lc=Pb$.
$$\left[ \begin{array}{ccc} 1.0 & 0.0 & 0.0 \\ 0.36 & 1.0 & 0.0 \\ 0.16 & 1.0 & 1.0 \end{array} \right] \left[ \begin{array}{c} c_1 \\ c_2 \\  c_3 \end{array} \right] = \left[ \begin{array}{ccc} 0.0 & 0.0 & 1.0 \\ 0.0 & 1.0 & 0.0 \\ 1.0 & 0.0 & 0.0 \end{array} \right] \left[ \begin{array}{c} 0.693147 \\ 1.098612 \\  1.386294 \end{array} \right] = \left[ \begin{array}{c} 1.386294 \\ 1.098612 \\   0.693147 \end{array} \right]$$

In [17]:
c_1 = 1.386294
c_2 = 1.098612 - .36*c_1
c_3 = 0.693147 - c_2 - .16*c_1
print("c_1 = %9.6e" %(c_1))
print("c_2 = %9.6e" %(c_2))
print("c_3 = %9.6e" %(c_3))

c_1 = 1.386294e+00
c_2 = 5.995462e-01
c_3 = -1.282062e-01


Hence we now have $Ux=c$.
$$U = \left[ \begin{array}{ccc} 25.0 & 5.0 & 1.0 \\ 0 & 1.2 & 0.64 \\ 0.0 & 0.0 & 0.2 \end{array} \right] \left[ \begin{array}{c} x_1 \\ x_2 \\  x_3 \end{array} \right] = \left[ \begin{array}{c} 1.386294 \\ 0.599546 \\  -0.128206 \end{array} \right] $$ 



In [18]:
x_3 = -0.128206/.2
x_2 = (0.599546 - .64*x_3)/1.2
x_1 = (1.386294-5*x_2-x_3)/25
print("x_1 = %9.6e" %(x_1))
print("x_2 = %9.6e" %(x_2))
print("x_3 = %9.6e" %(x_3))

x_1 = -8.720791e-02
x_2 = 8.415043e-01
x_3 = -6.410300e-01


### c)
Compute the residual norm $∥b − Ax∥_{\infty}$.

In [21]:
A = np.array([[4,2,1],[9,3,1],[25,5,1]])
x = np.array([[-8.720791e-02],[8.415043e-01],[-6.410300e-01]])
b = np.array([[0.693147],[1.098612],[1.386294]])
residual = b - np.dot(A,x)
print(residual)

[[  4.00000000e-08]
 [  2.90000000e-07]
 [  2.50000000e-07]]


Hence the residual norm is $2.9E^{-7}$

## Problem 2
Consider the task of interpolating a polynomial to the following three nodes $(x_i, y_i), i = 0,1,2$: 

$$(2.,0.693147), (3.,1.098612), (5.,1.386294)$$

(a) Construct the polynomial of lowest degree that fits these nodes, using the Lagrange formula. There is no need to expand out, but you could do that and use the result to check your results against what you get from polyfit.

(b) Use the above to compute an interpolated value at $x = 4.0$; by hand, not with polyval, but feel free to check your results with that.

(c) The above data are in fact given by $y_i = ln(x_i)$, so use the interpolation error formula to derive a numerical upper bound on the absolute interpolation errors at $x = 2.5$ and $x = 4.0$. Then compare the latter error bound to the actual error at $x = 4.0$.


### a)

$$P_2(x)=y_1\frac{(x-x_2)(x-x_3)}{(x_1-x_2)(x_1-x_3)}+y_2\frac{(x-x_1)(x-x_3)}{(x_2-x_1)(x_2-x_3)}+y_3\frac{(x-x_1)(x-x_2)}{(x_3-x_1)(x_3-x_2)}$$

$$P_2(x)=\frac{.693147(x-3)(x-5)}{(2-3)(2-5)}+\frac{1.098612(x-2)(x-5)}{(3-2)(3-5)}+\frac{1.386294(x-2)(x-3)}{(5-2)(5-3)} $$

In [4]:
print(.693147/((2-3)*(2-5)))

print(1.098612/((3-2)*(3-5)))

print(1.386294/((5-2)*(5-3)))

0.23104899999999998
-0.549306
0.23104899999999998


$$P_2(x)=.231049(x-3)(x-5)-.549306(x-2)(x-5)+.231049(x-2)(x-3) $$

### b) 

Now lets see what we get when we plug in x=4:

In [6]:
P_4 = .231049*(4-3)*(4-5)-.549306*(4-2)*(4-5)+.231049*(4-2)*(4-3) 
print(P_4)

1.3296609999999998


Hence $P_2(4) = 1.329661$

### c)
We need to determine the error for our polynomial...
$$f(x)-P_2(x) = \frac{(x-2)(x-3)(x-5)}{3!}f'''(c) $$
where $2 \leq c \leq 5 $
$$f(x)=ln(x) \rightarrow f'(x)=\frac{1}{x} \rightarrow f''(x)=-\frac{1}{x^2} \rightarrow f'''(x)=\frac{2}{x^3}$$

Hence to maximize the error, we need to maximize the third derivative. This can be done by letting $c = 2$.
Thus the error is,
$$e(x)=\frac{(x-2)(x-3)(x-5)}{3!}\frac{2}{2^3} $$

In [9]:
##For x=2.5
abs(((2.5-2)*(2.5-3)*(2.5-5)*2)/((6)*(2**3)))

0.026041666666666668

Hence the error  for $x=2.5$ is $0.026042$

In [10]:
##For x=4
abs(((4-2)*(4-3)*(4-5)*2)/((6)*(2**3)))

0.08333333333333333

Hence the error for $x=4$ is $0.08333333$

We can compare this maximum error to the actual error. 
The actual error is $\bigl\lvert  f(4)-P_2(4)\bigr\rvert$

In [14]:
log(4)-P_4

0.05663336111989081

Thus it is less than the upper bound of the error!

## 3
Verify that the one-sided three-point derivative approximation
$$f′(a) = \frac{−f(a+2h)+4f(a+h)−3f(a)}{2h}$$
is second order accurate:
$$f′(a) = \frac{−f(a + 2h) + 4f(a + h) − 3f(a)}{2h} + O(h2)$$

First lets calculate $f(x+h)$ and $f(x+2h)$.
$$f(x+2h) = f(x)+2hf'(x)+\frac{4}{2}h^2f''(x)+\frac{8}{6}h^3f'''(c_1) $$
$$f(x+h) = f(x)+hf'(x)+\frac{1}{2}h^2f''(x)+\frac{1}{6}h^3f'''(c_2) $$

Hence, 
$$f′(a) = \frac{−f(a+2h)+4f(a+h)−3f(a)}{2h} = \frac{1}{2h}(-f(x)-2hf'(x)-2h^2f''(x)-\frac{8}{6}h^3f'''(c_1)+4f(x)+4hf'(x)+2h^2f''(x)+\frac{4}{6}h^3f'''(c_2)-3f(x))$$

$$f'(x)+(\frac{1}{3}f'''(c_2)-\frac{2}{3}f'''(c_1)) $$

Hence we have a second order formula for $$f′(a) = \frac{−f(a+2h)+4f(a+h)−3f(a)}{2h}$$

#### 4
Verify that $f′(a) = \frac{−f(a+2h)+4f(a+h)−3f(a)}{2h}$ has degree of precision 2.

We are going to guess that $f'(x)=C_{-1}f(a+2h)+C_0f(a+h)+C_1f(a)$.

Now lets check that it works for n=0,n=1,n=2.

$\underline{n=0}: f(x)=1 \rightarrow f'(a) = 0 = C_{-1}+C_0+C_1$

$\underline{n=1}: f(x)=x-a \rightarrow f'(a) = 1 = 2hC_{-1}+C_0h $

$\underline{n=2}: f(x)=(x-a)^2 \rightarrow f'(a)=0 = 4h^2C_{-1}+C_0h^2 \rightarrow C_0=-4C_{-1} $
Hence after some solving we have,
$$C_{-1}=\frac{-1}{2h} $$
$$C_0=\frac{2}{h} $$
$$C_1=\frac{-3}{2h}$$

Now to check for $n=3$

$\underline{n=3}: f(a)=(x-a)^3 \rightarrow f'(a)=0=8h^3C_{-1}+h^3C_0$ 

This does not work with our previously found coefficients!

Therefore, $f′(a) = \frac{−f(a+2h)+4f(a+h)−3f(a)}{2h}$ has degree of precision 2.

## Problem 5
Apply Richardson extrapolation to the basic derivative approximation
$$f′(a) = \frac{f(a + h) − f(a)}{h} + O(h)$$
to get a higher order method.

$$F_n(x)=\frac{2^nF(h)-F(2h)}{2^n-1} $$
Since $n=1$ we have,
$$F_n(x)=\frac{2(f(a+h)-f(a)}{h}-\frac{f(a+2h)-f(a)}{2h} = \frac{4f(a+h)-f(a)-f(a+2h)+f(a)}{2h}=\frac{-f(a+2h)+4f(a+h)-3f(a)}{2h} $$

Thus we get a new method from richardson extrapolation. This is the same equation as from problem 3.